File: app/models/course.rb
class Course < ActiveRecord::Base
  HOLE_COUNTS = (9..27).to_a
  PENDING_STATUS = 'pending'
  APPROVED_STATUS = 'approved'
  STATUSES = [PENDING_STATUS, APPROVED_STATUS].freeze

  has_many :holes, dependent: :destroy
  has_many :rounds
  has_many :reviews, dependent: :destroy
  has_many :photos, through: :rounds
  belongs_to :submitter, class_name: "User"

  validates :state, :city, :country, :name, :status, presence: true
  validates_uniqueness_of :name, scope: :city
  validates :status, inclusion: STATUSES
  validates :state, inclusion: States.abbreviations

  scope :by_name, -> { order(:name) }
  scope :approved, -> { where(status: APPROVED_STATUS) }
  scope :available_to, ->(user) { where("courses.status = 'approved' OR courses.submitter_id = ?", user.id) }

  geocoded_by :full_street_address
  after_validation :geocode

  def full_street_address
    [address, city, state, country].compact.join(", ")
  end

  def rated?
    rating.present?
  end

  def hole_count
    holes.count
  end

  def hole_one
    holes.find_by(number: "1")
  end

  def scorecards
    Scorecard
      .joins(:round)
      .where(rounds: { course_id: id })
  end
end

File: app/assets/javascripts/models/course.js
App.Course = DS.Model.extend({
    city: DS.attr('string'),
    state: DS.attr('string'),
    country: DS.attr('string'),
    name: DS.attr('string'),
    status: DS.attr('string'),
    location: DS.attr('string'),
    roundsPlayed: DS.attr('number'),
    imageUrl: DS.attr('string'),
    scorecards: DS.hasMany('scorecard', { async: true }),
    holes: DS.hasMany('hole', { async: true }),
    holeCount: DS.attr('number'),

    isApproved: function() {
        return this.get('status') === 'approved';
    }.property('status'),

    hasImage: function() {
        return !Ember.isEmpty(this.get('imageUrl'));
    }.property('imageUrl'),

    imageUploadPath: function() {
      return "/api/courses/" + this.get('id') + "/course_images";
    }.property('id')
});

App.AvailableCourse = App.Course.extend();
